### PR TITLE
Allow file creation without providing a filename in User-Metadata header

### PR DESCRIPTION
**Pull request by alucic**
Even if the client doesn't send `User-Metadata` keys `name` and/or `filename` server should still be able to proceed with handling `POST` request.

In that case `uploadKey` UUID is used as a filename. 

More info can be found here
https://tus.io/protocols/resumable-upload.html#how-can-i-get-the-file-name-or-file-type-for-an-upload

If I'm reading it correctly, `tusd` implementation completely ignores `User-Metadata` header when creating a new file:
https://github.com/tus/tusd/blob/9c0d753f7dddedc1444178171add376cea91ec30/filestore/filestore.go#L62-L66

I think this is a nice feature and it can't hurt, 1 less `400 Bad request` to worry about

### CODE REVIEW

File: tests/Tus/ServerTest.php [hunk 1]
```diff
@@ -714,28 +714,6 @@ public function it_handles_head_request_for_final_upload()
...
-                'REQUEST_METHOD' => 'POST',
-                'REQUEST_URI' => '/files',
-            ]);
-
-        $response = $this->tusServerMock->handlePost();
-
-        $this->assertEquals(400, $response->getStatusCode());
-        $this->assertEmpty($response->getContent());
-    }
-
```
Comment (samundra) [inline tests/Tus/ServerTest.php]: Why was this removed ?

Comment (alucic) [reply tests/Tus/ServerTest.php]: 400 was returned if the filename was missing. With this PR, sending filename is optional because there's a fallback, this test is not needed anymore.

https://github.com/ankitpokhrel/tus-php/blob/3878a4933d9e62d4e9a543e88d2b76fd4787d0e1/src/Tus/Server.php#L347-L349

Comment (samundra) [reply tests/Tus/ServerTest.php]: But if we remove this test then our test coverage will drop right? I think it's better to keep test intact even if fallback covers it. Or better remove this code as well. @ankitpokhrel What do you think?

Comment (ankitpokhrel) [reply tests/Tus/ServerTest.php]: @alucic we can remove `if (empty($fileName)) { ` check since that piece of code is not reachable anymore and it is better to add a test to make sure we are getting auto generated filename if it is not provided in the meta.

Comment (alucic) [reply tests/Tus/ServerTest.php]: Isn't that covered with `it_extracts_file_name_has_fallback`? Am I missing something?

File: src/Request.php [hunk 2]
```diff
@@ -112,13 +113,22 @@ public function extractFromHeader(string $key, string $value) : array
...
-        return $this->extractMeta('name') ?: $this->extractMeta('filename');
+        if ($this->extractMeta('name')) {
+            return $this->extractMeta('name');
+        }
+
+        if ($this->extractMeta('filename')) {
+            return $this->extractMeta('filename');
+        }
+
+        return Uuid::uuid4()->toString();
```
Comment (ankitpokhrel) [inline src/Request.php]: We can shorten this method

```php
public function extractFileName() : string
{
    $filename = $this->extractMeta('name') ?: $this->extractMeta('filename');

    return ! empty($filename) ? $filename : Uuid::uuid4()->toString();
}
```

File: tests/Tus/ServerTest.php (hunk 1 above)
Comment (ankitpokhrel) [reply tests/Tus/ServerTest.php]: I would also add one test it in `ServerTest.php` by not passing `Upload-Metadata` header, it should take the `UUID` in that case.

**Comment by ankitpokhrel** [comment]
Instead of generating new uuid we can actually use upload key as a filename if the filename is not present in the upload meta. This will require minimal change to achieve what we want.

```diff
diff --git a/src/Tus/Server.php b/src/Tus/Server.php
index c6391e0..822cac6 100644
--- a/src/Tus/Server.php
+++ b/src/Tus/Server.php
@@ -344,16 +344,17 @@ class Server extends AbstractTus
         $fileName   = $this->getRequest()->extractFileName();
         $uploadType = self::UPLOAD_TYPE_NORMAL;

-        if (empty($fileName)) {
-            return $this->response->send(null, HttpResponse::HTTP_BAD_REQUEST);
-        }
-
         if ( ! $this->verifyUploadSize()) {
             return $this->response->send(null, HttpResponse::HTTP_REQUEST_ENTITY_TOO_LARGE);
         }

         $uploadKey = $this->getUploadKey();
-        $filePath  = $this->uploadDir . DIRECTORY_SEPARATOR . $fileName;
+
+        if (empty($fileName)) {
+            $fileName = $uploadKey;
+        }
+
+        $filePath = $this->uploadDir . DIRECTORY_SEPARATOR . $fileName;

         if ($this->getRequest()->isFinal()) {
             return $this->handleConcatenation($fileName, $filePath);
```

@alucic @samundra any thoughts on this?

File: tests/Tus/ServerTest.php (hunk 1 above)
Comment (alucic) [reply tests/Tus/ServerTest.php]: Oh, I just reread the comments and figured why there was a confusion, I didn't remove that it if statement that returns 400.

I'll update the PR to use uploadKey instead generating a new UUID for the filename

**Comment by alucic** [comment]
@ankitpokhrel 
I like the approach of using `$uploadKey` instead of generating a new UUID.
I believe this PR is ready for another round of CR.

I'm a bit confused about `Upload-Key` header. That header doesn't seem to be a part of TUS protocol but I'll open a new issue where we can discuss it. I've seen it being used in the `Client` but I'm not 100% sure that we need it.

Thanks!

File: tests/Tus/ServerTest.php [hunk 3]
```diff
@@ -1077,6 +1055,231 @@ public function it_handles_post_request()
         $this->assertEquals('1.0.0', $response->headers->get('tus-resumable'));
     }
 
+    /**
+     * @test
+     *
+     * @covers ::handlePost
+     * @covers \TusPhp\Events\UploadCreated::__construct
+     */
+    public function it_handles_post_request_uses_uploadkey_when_metadata_is_not_present()
```
Comment (ankitpokhrel) [inline tests/Tus/ServerTest.php]:
```suggestion
    public function it_uses_uploadkey_as_filename_when_metadata_is_not_present()
```

File: tests/Tus/ServerTest.php [hunk 4]
```diff
@@ -1077,6 +1055,231 @@ public function it_handles_post_request()
...
+        $this->assertEquals('1.0.0', $response->headers->get('tus-resumable'));
+    }
+
+    /**
+     * @test
+     *
+     * @covers ::handlePost
+     * @covers \TusPhp\Events\UploadCreated::__construct
+     */
+    public function it_handles_post_request_uses_uploadkey_when_metadata_name_is_empty_string()
```
Comment (ankitpokhrel) [inline tests/Tus/ServerTest.php]:
```suggestion
    public function it_uses_uploadkey_as_filename_when_metadata_name_is_empty_string()
```

**Comment by alucic** [comment]
I'll get to this today
